examples/README.md

```diff
@@ -0,0 +1,46 @@
+# RADIO Examples
+
+## K-NN Classification
+
+K-NN classification consists of embedding all of the summary vectors produced by the model into a database. We then can classify a new vector given this database by computing a weighted sum of the $k$ nearest vectors in the database.
+
+We follow the algorithm in [Unsupervised Feature Learning via Non-Parametric Instance Discrimination](https://arxiv.org/pdf/1805.01978.pdf) (Section 3.4) for the weighting. Let $\mathbf{q} \in \mathbb{R}^C$ be the query vector, and $\mathbf{D} \in \mathbb{R}^{N \times C}$ with $N$ being the size of the training set, and $C$ the dimension of the summary vectors. Let $\mathbf{d}_i$ be the $i$-th row of $\mathbf{D}$, let $s_i = \cos(\mathbf{d}_i, \mathbf{q})$ be the cosine similarity between the database vector and the query vector, and let $c_i$ be the class corresponding to vector $\mathbf{d}_i$.
+
+The top $k$ nearest neighbors, denoted by $\mathcal{N}_k$, is the set of vectors of size $k$ with highest cosine similarity over $\mathbf{D}$. Then, class $c$ gets a total weight
+$$
+\begin{aligned}
+    \alpha_i &= e^{s_i / \tau} \\
+    w_c &= \sum_{i \in \mathcal{N}_k} \mathbb{1}\left(c_i = c\right) \cdot \alpha_i \\
+    c_{\mathbf{q}} &= \argmax_c w_c
+\end{aligned}
+$$
+
+Following Wu et.al. we set $\tau = 0.07$, thus $c_{\mathbf{q}}$ is the predicted class. Note that this is the same formulation that [DINO](http://arxiv.org/abs/2104.14294) and [DINOv2](http://arxiv.org/abs/2304.07193) use.
+
+We argue that K-NN classification is a great evaluation protocol for holistic image tasks because it works for any vision model that can produce a summary vector for an image, and thus it easily allows us to compare, for example, CLIP vs DINOv2 vs SAM vs etc. The technique is not zero-shot like CLIP in the sense that it still requires a scan through the training set, however, it's also a training-free technique wrt the model.
+
+### Evaluation Script
+
+We provide an implementation of the K-NN algorithm with with the `knn_classification.py` script, which uses the HuggingFace Datasets library to make evaluation simple. This script works in both single-GPU and multi-GPU setups. For multi-GPU, replace `python` with `torchrun`.
+
```

Comment:
maybe we could add an example multi-GPU invocation to the README?

Reply:
I'm a bit worried that it's environment dependent. I'm not sure if the specific way that I'm running it on our cluster is a general setup, or something that we're doing at NVIDIA. The docs for `torchrun` seem to suggest that  it "just works" generally.

Reply:
Yes I find torchrun to be more portable than mpirun :-) - In single node setups, I use `--nproc_per_node` to specify the number of GPUs.

Reply:
Is `--nproc_per_node` necessary? Or if I just launch with no settings, would that pick up all GPUs on the system?

Reply:
I think `--nproc-per-node` is necessary if nothing else sets up the environment variables for distributed execution.

Reply:
So `torchrun --nproc-per-node 8 knn_classification.py ...` is likely a sufficient command?

Reply:
Yes is my go-to incantation :-)

Reply:
Cool. Done